confluent_kafka/src/confluent_kafka.c

```diff
@@ -561,6 +584,10 @@ PyObject *Message_new0 (const Handle *handle, const rd_kafka_message_t *rkm) {
 
 	self->timestamp = rd_kafka_message_timestamp(rkm, &self->tstype);
 
+    if (!rd_kafka_message_headers(rkm, &hdrs)) {
```

Comment:
Object creation is the most expensive thing in Python so I think we should put this off until headers are actually requested, in the headers() accessor method.

So I suggest we just extract the C headers list here and keep it on the Message object, making sure to call set_headers(NULL) on the message to transfer ownership to the Python code.